### PR TITLE
Correct saving files to DropBox/Drive/OneDrive

### LINKED ISSUE #197: Saving DB fails intermittently on windows
## Current Behavior
Most of the time, attempting to save a database file on windows fails with this message:

    Writing the database failed.
    
    The Process cannot access the file because it is being used by another process.

A copy of the DB file is written at \path\to\db.kdbx.<2 random? letters>\<PID>

Occasionally, the write does go through without an error.

## Steps to Reproduce (for bugs)
I have not yet been able to reproduce the error with a fresh database created with either keepassxc or keepassX. Obviously, I'm not willing to upload one of my real DBs.

## Context
Process Explorer reports no other process has a handle to the .kdbx file.

## Your Environment
* KeePassXC version/commit used: 42c0815cdaa652bcdf25b90a2a4a20b2981a39b2
* Qt version 5.6.2
* Compiler: gcc version 6.3.0 (Rev1, Built by MSYS2 project)
* Operating System and version: Windows 7 SP1

### LINKED ISSUE #1159: Auto backup database on save
## Expected Behavior

Create a backup of opened every time a database with changes is being saved.

## Current Behavior

No backup solution.

## Possible Solution

New application settings to automatically create backups similar to [SimpleDatabaseBackup](https://github.com/jnko/SimpleDatabaseBackup) or [DataBaseBackup](https://keepass.info/plugins.html#databasebackup) with maximum backup copies and path to backups (and possible remote locations FTP/Samba/WebDav etc...).

With identified names by date <db_name>_YYYY-MM-DD-HH-mm.kdbx

## Context

Backups are useful in case of corrupted database or undo recent changes, having an option to create backups on changes can avoid future issues with databases.

### PR DESCRIPTION
**Pull request by droidmonkey**
## Description

* Added backup before save config setting
* Added non-atomic save config setting

* Adds prompt to fallback to non-atomic saves if saving fails 3 times:

failed_saves

* Updated basic settings to be in logical groups:

new_settings

Fixes #197 and closes #1159 

## Motivation and context

The risk of data loss is minimal and better than not being able to save at all.

## How has this been tested?

Manually and automatically

## Types of changes

- ✅ Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- ✅ I have read the **CONTRIBUTING** document. **[REQUIRED]**
- ✅ My code follows the code style of this project. **[REQUIRED]**
- ✅ All new and existing tests passed. **[REQUIRED]**
- ✅ I have compiled and verified my code with `-DWITH_ASAN=ON`. **[REQUIRED]**

### CODE REVIEW

File: src/core/Database.cpp [hunk 1]
```diff
@@ -470,30 +470,52 @@ Database* Database::unlockFromStdin(QString databaseFilename, QString keyFilenam
...
         if (writer.hasError()) {
+            // the writer failed
             return writer.errorString();
         }
 
-        if (saveFile.commit()) {
+        saveFile.close(); // flush to disk
+
+        if (keepOld) {
+            QFile::remove(filePath + ".old");
```
Comment (phoerious) [inline src/core/Database.cpp]: .old should be inserted between the base filename and the .kdbx extension.

File: src/core/Database.cpp [hunk 2]
```diff
@@ -470,30 +470,52 @@ Database* Database::unlockFromStdin(QString databaseFilename, QString keyFilenam
...
+            QFile::rename(filePath, filePath + ".old");
+        }
+
+        QFile::remove(filePath);
+        if (saveFile.rename(filePath)) {
```
Comment (phoerious) [inline src/core/Database.cpp]: Are you sure this is going to work? If the temp folder is on a different partition, then renaming a file means copying it. Given this is rare on Windows, I'd still see it as a potential problem, because corruption could occur during copying.
Another problem I see is that on platforms other than Windows, the temp location is often /tmp, which is world-readable (the file may not be, but you can still at least list it). The database is encrypted, but we still shouldn't move it to such a location without the user's consent. Maybe make the QTemporaryFile fix Windows-only and continue using QSaveFile on other platforms. Outside of Dropbox folders it's by far the safer option and this problem is Windows-only.

File: src/core/Database.cpp (hunk 1 above)
Comment (droidmonkey) [reply src/core/Database.cpp]: Why you make this difficult? 😝

File: src/core/Database.cpp (hunk 2 above)
Comment (droidmonkey) [reply src/core/Database.cpp]: Yah I got some education from that bug thread, good to know. I did consider the option of only implementing the temp file scheme in windows.

Another issue is if a database is saved to a non-writable directory, but the database itself is writable, this implementation will actually delete the database and return.

File: src/core/Database.cpp (hunk 1 above)
Comment (phoerious) [reply src/core/Database.cpp]: 😏

File: src/core/Database.cpp (hunk 2 above)
Comment (phoerious) [reply src/core/Database.cpp]: That's indeed a major problem.

**Comment by droidmonkey** [comment]
This PR is ready for final review & merge.

File: src/gui/DatabaseTabWidget.cpp [hunk 3]
```diff
@@ -322,12 +324,15 @@ bool DatabaseTabWidget::saveDatabase(Database* db, QString filePath)
         }
 
         dbStruct.dbWidget->blockAutoReload(true);
-        QString errorMessage = db->saveToFile(filePath);
+        // TODO: Make this async, but lock out the database widget to prevent re-entrance
+        bool useAtomicSaves = config()->get("UseAtomicSaves").toBool();
+        QString errorMessage = db->saveToFile(filePath, useAtomicSaves, config()->get("BackupBeforeSave").toBool());
```
Comment (phoerious) [inline src/gui/DatabaseTabWidget.cpp]: At least the UseAtomicSaves settings should have true as default value in Config::init().

File: src/gui/DatabaseTabWidget.cpp [hunk 4]
```diff
@@ -336,6 +341,22 @@ bool DatabaseTabWidget::saveDatabase(Database* db, QString filePath)
         } else {
             dbStruct.modified = true;
             updateTabName(db);
+
+            if (++dbStruct.saveAttempts > 2 && useAtomicSaves) {
```
Comment (phoerious) [inline src/gui/DatabaseTabWidget.cpp]: I would check for write permissions first and show a dedicated error message. Otherwise the user will disable atomic saves for nothing if the problem is actually missing write permissions.
I know this is an edge case, because without write permissions, the DB will be opened in read-only mode, but it can still occur, if write permissions are somehow removed mid-air.

File: src/gui/SettingsWidgetGeneral.ui [hunk 5]
```diff
@@ -34,124 +34,134 @@
...
-          <bool>true</bool>
+        <widget class="QGroupBox" name="saveGroup">
+         <property name="title">
+          <string>File Management</string>
          </property>
+         <layout class="QVBoxLayout" name="verticalLayout_5">
+          <item>
+           <widget class="QCheckBox" name="useAtomicSavesCheckBox">
+            <property name="text">
+             <string>Safely save database files (may be incompatible with DropBox, etc)</string>
```
Comment (phoerious) [inline src/gui/SettingsWidgetGeneral.ui]: Dropbox's official spelling is with lowercase b.

File: src/core/Config.cpp [hunk 6]
```diff
@@ -114,6 +114,8 @@ void Config::init(const QString& fileName)
     m_defaults.insert("AutoSaveAfterEveryChange", false);
```
[the remark this answers is not in the document]
Comment (droidmonkey) [reply src/core/Config.cpp]: Yes good idea

File: src/gui/DatabaseTabWidget.cpp (hunk 4 above)
Comment (droidmonkey) [reply src/gui/DatabaseTabWidget.cpp]: You can't easily check for write conditions on windows. I did a fair bit of research on the matter while trying to diagnose.

File: src/gui/DatabaseTabWidget.cpp (hunk 3 above)
Comment (droidmonkey) [reply src/gui/DatabaseTabWidget.cpp]: It does

Comment (phoerious) [reply src/gui/DatabaseTabWidget.cpp]: Yeah, forget what I said. 😕

File: src/gui/DatabaseTabWidget.cpp (hunk 4 above)
Comment (phoerious) [reply src/gui/DatabaseTabWidget.cpp]: Hmm. But that would also mean that the QTemporaryFile workaround could just delete the file when it's unable to move the temporary file to the original location after deleting the old one. I don't know how it's on Windows, but when you own a file, you can always delete it, but you may not be able to create a new one in the same directory.

Comment (phoerious) [reply src/gui/DatabaseTabWidget.cpp]: I have an idea: check the return value of `tempFile.rename()` when moving the temporary file and show a save file dialog to choose a new location as long as it's returning false.

Comment (phoerious) [reply src/gui/DatabaseTabWidget.cpp]: OK, leave as is then.

File: src/gui/DatabaseTabWidget.cpp [hunk 7]
```diff
@@ -336,6 +341,22 @@ bool DatabaseTabWidget::saveDatabase(Database* db, QString filePath)
...
+                // Saving failed 3 times, issue a warning and attempt to resolve
+                auto choice = MessageBox::question(this, tr("Disable safe saves?"),
+                                                   tr("KeePassXC has failed to save the database multiple times. "
+                                                   "This is likely caused by file sync services holding a lock on "
+                                                   "the save file.\nDisable safe saves and try again?"),
```
Comment (louib) [inline src/gui/DatabaseTabWidget.cpp]: @droidmonkey shouldn't we also use the `atomic saves` terminology here?

Comment (phoerious) [reply src/gui/DatabaseTabWidget.cpp]: I was thinking that, too. But I don't know how well users will understand it. Atomic in a non-nuklear context is a very domain-specific word.

Comment (droidmonkey) [reply src/gui/DatabaseTabWidget.cpp]: Atomic is a developer term. I wanted to dumb it down on the user interface.

Comment (calmarc) [reply src/gui/DatabaseTabWidget.cpp]: User (me) googled it: "write to tmp-file and then (atomic->) move over". (had no idea before searching for the term - sounded a little intimidating to me :) But some little hint texts, should do the trick - I guess.
